include/glaze/ext/jsonrpc.hpp

```diff
@@ -0,0 +1,538 @@
+         [](auto const&, auto const&) {}};
+   };
+   namespace concepts
+   {
+      template <typename method_t>
+      concept method_type = requires(method_t meth) {
+                               method_t::name_v;
+                               {
+                                  std::same_as<decltype(method_t::name_v), std::string_view>
+                               };
+                               typename method_t::params_t;
+                               typename method_t::jsonrpc_id_t;
+                               typename method_t::request_t;
+                               typename method_t::response_t;
+                               {
+                                  meth.callback
+                               };
+                            };
+
+      template <typename method_t>
+      concept server_method_type =
+         requires(method_t meth) {
+            requires method_type<method_t>;
+            requires std::same_as<
+               decltype(meth.callback),
+               std::function<expected<typename method_t::result_t, rpc::error>(typename method_t::params_t const&)>>;
+         };
+
+      template <typename method_t>
+      concept client_method_type =
+         requires(method_t meth) {
+            requires method_type<method_t>;
+            requires std::same_as<decltype(meth.callback),
+                                  std::function<void(glz::expected<typename method_t::result_t, rpc::error> const&,
+                                                     jsonrpc_id_type const&)>>;
+         };
+   }
+
+   namespace detail
+   {
+      template <detail::basic_fixed_string name, typename... method_type>
+      inline constexpr void set_callback(glz::tuplet::tuple<method_type...>& methods, auto const& callback)
+      {
+         constexpr bool method_found = ((method_type::name_v == name) || ...);
+         static_assert(method_found, "Method not settable in given tuple.");
+
+         methods.any([&callback](auto&& method) -> bool {
+            if constexpr (name == std::remove_reference_t<decltype(method)>::name_v) {
+               method.callback = callback;
+               return true;
+            }
+            return false;
+         });
+      }
+   }
+
+   template <concepts::server_method_type... method_type>
+   struct server
+   {
+      server() = default;
+      glz::tuplet::tuple<method_type...> methods{};
+
+      template <detail::basic_fixed_string name>
+      constexpr void on(auto const& callback)  // std::function<expected<result_t, rpc::error>(params_t const&)>
+      {
+         detail::set_callback<name>(methods, callback);
+      }
+
+      // Return json stringified response_t meaning it can both be error and non-error
+      // If `id` is null in json_request a response will not be generated
+      std::vector<std::pair<std::string, rpc::error>> call(std::string_view json_request)
```

Comment:
This is really mind boggling for me, I am not sure what is the best return type here.

But yes a complete response string would be much easier, but what I would also like to accomplish is the one to one relation between the response of one request and its error than a user can act on individual errors.

Maybe something like this would be appropriate here:
```C++
struct server {
  struct response {
    std::vector<std::string> responses{};
    std::vector<rpc::error> errors{};
    bool empty() const noexcept;
    std::string complete_response() const; // Ideally if we could concat responses with [responses[0], ... responses[N]] as std::string_view
  };
  
  response call(std::string_view json_request);
};
```

Reply:
In the case you want to handle errors individually I would return would std::vector<response_t<glz::raw_json>> since the response_t contains the rpc::error and you can call write_json on the entire std::vector<response_t<glz::raw_json>> to get the string.

Reply:
I think this is a good idea, something like this:

```C++
template<typename return_t = std::string>
return_t call(std::string_view) {
if constexpr(std::same_as<return_t, std::string>) {
return std::string{}; // complete response for batch request or request
}
else if constexpr(std::same_as<return_t, std::vector<rpc::generic_response_t>>) {
return std::vector<rpc::generic_response_t>{}; // serializable return and accessible error
}
else {
static_assert(false); // I know this does not work but just to show the example
}
}

void usage() {
std::string response = server.call(""); // [{},{},...]
auto response_v = server.call<std::vector<rpc::generic_response_t>>("");
std::string responsestr = glz::write_json(response_v);
}

```